cocotb/types/__init__.py

```diff
@@ -0,0 +1,43 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+from .range import Range  # noqa: F401
+from .array import Array  # noqa: F401
+
+
+def concat(a: Array, b: Array) -> Array:
+    """
+    Create a new array that is the concatenation of one array with another.
+
+    Uses the :meth:`__concat__` or :meth:`__rconcat__` special methods to dispatch to a particular implementation,
+    exactly like other binary operations in Python.
+
+    Raises:
+        TypeError: when the arguments do not support concatenation in the given order.
+    """
+    a_concat = getattr(a.__class__, "__concat__", None)
+    a_rconcat = getattr(a.__class__, "__rconcat__", None)
+    b_rconcat = getattr(b.__class__, "__rconcat__", None)
+
+    if isinstance(b, a.__class__) and a_rconcat != b_rconcat:
```

Comment:
```suggestion
    if issubclass(type(b), type(a)) and type(a) is not type(b) and a_rconcat is not b_rconcat:
```

But I'd recommend you just take the `desugar` version with less editorialization.

Reply:
I've read that using `is not` for type comparisons is not advised, I'm guessing that's because parameterized types (depending on the implementation) might not return the same object given the same parameters, type `==` can handle that.

Using `isinstance` is equivalent here and I don't get why there needs to be protection against matching the same type, seeing as the second part of the condition will not match in that case. If people write code in a non-intuitive way and *don't* put a comment as to why they did it that way, I consider it superfluous.

Reply:
> I've read that using is not for type comparisons is not advised,

I think that's usually in the context of "are you sure you didn't mean `not issubclass(X, Y)`"

> If people write code in a non-intuitive way and don't put a comment as to why they did it that way, I consider it superfluous.

The comment is in the description of the package as a whole - it's written to exactly match what CPython does internally.

CPython does things the way it does for historical reasons; in particular, the `a_rconcat != b_rconcat` condition was a mistake that was too late to revert.
If we're going to add a python-like dunder hook, then we should make it python like, along with all the quirks. Otherwise, the absence of those quirks can bite users.

> parameterized types (depending on the implementation) might not return the same object given the same parameters

IMO this would be a very bad design.
At any rate, Cpython internall uses `is` not `==`, so we should match it.

Reply:
[CPython binary operation for Python-defined types](https://github.com/python/cpython/blob/a328d73843cfd42d2aee1434c78df1ef2845931a/Objects/typeobject.c#L6940).

Reply:
> the a_rconcat != b_rconcat condition was a mistake that was too late to revert.

The mistake is to check that at all, right? If the right hand side is a subclass of the left, the right `__rop__` should *always* be called first (if it exists)?